core/container/avl/avl.odin

```diff
@@ -0,0 +1,598 @@
+/*
+package avl implements an AVL tree.
+
+The implementation is non-intrusive, and non-recursive.
+*/
+package container_avl
+
+import "base:runtime"
+_ :: runtime
+
+// Originally based on the CC0 implementation by Eric Biggers
+// See: https://github.com/ebiggers/avl_tree/
+
+// Direction specifies the traversal direction for a tree iterator.
+Direction :: enum i8 {
+	// Backward is the in-order backwards direction.
+	Backward = -1,
+	// Forward is the in-order forwards direction.
+	Forward  = 1,
+}
+
+// Tree is an AVL tree.
+Tree :: struct($Value: typeid) {
+	on_remove:           proc(value: Value, user_data: rawptr),
+	on_remove_user_data: rawptr,
+
+	_root:               ^Node(Value),
+	_node_allocator:     runtime.Allocator,
+	_cmp_fn:             proc(a, b: Value) -> int,
+	_size:               int,
+}
+
+// Node is an AVL tree node.
+//
+// WARNING: It is unsafe to mutate value if the node is part of a tree
+// if doing so will alter the Node's sort position relative to other
+// elements in the tree.
+Node :: struct($Value: typeid) {
+	value:    Value,
+
+	_parent:  ^Node(Value),
+	_left:    ^Node(Value),
+	_right:   ^Node(Value),
+	_balance: i8,
+}
+
+// Iterator is a tree iterator.
+//
+// WARNING: It is unsafe to modify the tree while iterating, except via
+// the iterator_remove method.
+Iterator :: struct($Value: typeid) {
+	_tree:        ^Tree(Value),
+	_cur:         ^Node(Value),
+	_next:        ^Node(Value),
+	_direction:   Direction,
+	_initialized: bool,
+}
+
+// init initializes a tree.  The comparision function MUST return
+// < 0, 0, or > 0 if a < b, a == b, or a > b respectively.
+init :: proc(
+	t: ^Tree($Value),
+	cmp_fn: proc(a, b: Value) -> int,
+	node_allocator := context.allocator,
+) {
+	t._root = nil
+	t._node_allocator = node_allocator
+	t._cmp_fn = cmp_fn
+	t._size = 0
+}
+
+// destroy de-initializes a tree.
+destroy :: proc(t: ^Tree($Value), call_on_remove: bool = true) {
+	iter := iterator(t, Direction.Forward) // TODO: Post-order would be better.
+	for node := iterator_first(&iter); node != nil; node = iterator_next(&iter) {
```

Comment:
When you make `iterator_next` return a `bool` and make `iterator` initialize itself, you can just make this `for node := iterator_next(&iter) {`